pysparkle/pysparkle/etl.py

```diff
@@ -0,0 +1,141 @@
+    spark: SparkSession,
+    container_name: str,
+    datasource_path: str,
+    datasource_type: str,
+    spark_read_options: Optional[dict[str, Any]] = None,
+) -> DataFrame:
+    """Reads the most recent data, based on rundate, from an ADLS location and returns it as a dataframe.
+
+    The function identifies the latest rundate by parsing directory names in the given path and then reads the
+    corresponding data. The rundate directories should follow the format: "rundate=<date_string_in_ISO-8601>",
+    e.g. "rundate=YYYY-MM-DDTHH:MM:SS" or "rundate=YYYY-MM-DDTHHMMSS.FFFFFFZ".
+
+    Args:
+        spark: Spark session.
+        container_name: Name of the ADLS container.
+        datasource_path: Directory path within the ADLS container where rundate directories are located.
+        datasource_type: Source system type that Spark can read from, e.g. delta, table, parquet, json, csv.
+        spark_read_options: Options to pass to the DataFrameReader.
+
+    Returns:
+        The dataframe loaded from the datasource with the most recent rundate, with metadata columns dropped.
+
+    """
+    logger.info(f"Reading dataset: {datasource_path}")
+    dirname_prefix = "rundate="
+    metadata_columns = ["meta_ingestion_datetime", "meta_ingestion_pipeline", "meta_ingestion_run_id"]
+    directories = get_adls_directory_contents(container_name, datasource_path, recursive=False)
+    rundates = [directory.split(dirname_prefix)[1] for directory in directories]
+    most_recent_rundate = max(rundates, key=isoparse)
+    logger.info(f"Latest rundate: {most_recent_rundate}")
+    latest_path = Path(datasource_path) / (dirname_prefix + most_recent_rundate)
+    dataset_url = get_adls_file_url(container_name, str(latest_path))
+    return read_datasource(spark, dataset_url, datasource_type, spark_read_options).drop(*metadata_columns)
+
+
+def transform_and_save_as_delta(
+    spark: SparkSession,
+    input_df: DataFrame,
+    transform_func: Callable[[DataFrame], DataFrame],
+    target_container: str,
+    output_file_name: str,
+    overwrite: bool = True,
+    merge_keys: Optional[list[str]] = None,
+) -> None:
+    """Transforms an input dataframe using the provided transformation function and saves the result as a Delta table.
+
+    Args:
+        spark: Spark session.
+        input_df: Data frame to be transformed.
+        transform_func: Transformation function.
+        target_container: Name of the destination container in ADLS.
+        output_file_name: The name of the file (including any subdirectories within the container).
+        overwrite: Flag to determine whether to overwrite the entire table or perform an upsert.
+        merge_keys: List of keys based on which upsert will be performed.
+
+    """
+    transformed_df = transform_func(input_df)
+    output_filepath = get_adls_file_url(target_container, output_file_name)
+    save_dataframe_as_delta(spark, transformed_df, output_filepath, overwrite, merge_keys)
```

Comment:
We should add some metadata columns to the output table, similar to the ones in the ingest table.
(Happy for this to be a separate task / PR)

Reply:
Task created: https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6724

Reply:
Added here - https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/6724